File: solutions/python/is_binary_tree_a_bst_const_space.py
# is_binary_tree_a_BST_const_space.cpp 4a4c5f91493e5e482eaa79892816c1ccefa084f4
from binary_tree_prototype import BinaryTreeNode


# @include
def is_binary_tree_BST(root):
    n = root
    # Stores the value of previous visited node.
    prev_value = float('-inf')
    result = True

    while n:
        if n.left:
            # Finds the predecessor of n.
            pre = n.left
            while pre.right and pre.right is not n:
                pre = pre.right

            # Processes the successor link.
            if pre.right:  # pre.right is n.
                # Reverts the successor link if predecessor's successor is n.
                pre.right = None
                if prev_value > n.data:
                    result = False
                prev_value = n.data
                n = n.right
            else:  # If predecessor's successor is not n.
                pre.right = n
                n = n.left
        else:
            if prev_value > n.data:
                result = False
            prev_value = n.data
            n = n.right
    return result
# @exclude


def main():
    #      3
    #    2   5
    #  1    4 6
    tree = BinaryTreeNode(3)
    tree.left = BinaryTreeNode(2)
    tree.left.left = BinaryTreeNode(1)
    tree.right = BinaryTreeNode(5)
    tree.right.left = BinaryTreeNode(4)
    tree.right.right = BinaryTreeNode(6)
    # should output True.
    assert is_binary_tree_BST(tree) == True
    print(is_binary_tree_BST(tree))
    #      10
    #    2   5
    #  1    4 6
    tree.data = 10
    # should output False.
    assert not is_binary_tree_BST(tree)
    print(is_binary_tree_BST(tree))
    # should output True.
    assert is_binary_tree_BST(None) == True
    print(is_binary_tree_BST(None))


if __name__ == '__main__':
    main()

File: solutions/cpp/is_binary_tree_a_bst_const_space.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <cassert>
#include <iostream>
#include <limits>
#include <memory>

#include "./Binary_tree_prototype.h"

using std::boolalpha;
using std::cout;
using std::endl;
using std::make_unique;
using std::numeric_limits;
using std::unique_ptr;

// @include
bool IsBinaryTreeBST(const unique_ptr<BinaryTreeNode<int>>& root) {
  auto* n = root.get();
  // Stores the value of previous visited node.
  int prevValue = numeric_limits<int>::min();
  bool result = true;

  while (n) {
    if (n->left.get()) {
      // Finds the predecessor of n.
      auto* pre = n->left.get();
      while (pre->right.get() && pre->right.get() != n) {
        pre = pre->right.get();
      }

      // Processes the successor link.
      if (pre->right.get()) {  // pre->right == n.
        // Reverts the successor link if predecessor's successor is n.
        pre->right.release();
        if (prevValue > n->data) {
          result = false;
        }
        prevValue = n->data;
        n = n->right.get();
      } else {  // If predecessor's successor is not n.
        pre->right.reset(n);
        n = n->left.get();
      }
    } else {
      if (prevValue > n->data) {
        result = false;
      }
      prevValue = n->data;
      n = n->right.get();
    }
  }
  return result;
}
// @exclude

int main(int argc, char* argv[]) {
  //      3
  //    2   5
  //  1    4 6
  auto root = make_unique<BinaryTreeNode<int>>(BinaryTreeNode<int>{3});
  root->left = make_unique<BinaryTreeNode<int>>(BinaryTreeNode<int>{2});
  root->left->left = make_unique<BinaryTreeNode<int>>(BinaryTreeNode<int>{1});
  root->right = make_unique<BinaryTreeNode<int>>(BinaryTreeNode<int>{5});
  root->right->left =
      make_unique<BinaryTreeNode<int>>(BinaryTreeNode<int>{4});
  root->right->right =
      make_unique<BinaryTreeNode<int>>(BinaryTreeNode<int>{6});
  assert(IsBinaryTreeBST(root) == true);
  cout << boolalpha << IsBinaryTreeBST(root) << endl;
  //      10
  //    2   5
  //  1    4 6
  root->data = 10;
  // should output false
  assert(!IsBinaryTreeBST(root));
  cout << boolalpha << IsBinaryTreeBST(root) << endl;
  // should output true
  assert(IsBinaryTreeBST(nullptr) == true);
  cout << boolalpha << IsBinaryTreeBST(nullptr) << endl;
  return 0;
}
